File: cp5200/cp5200_test.py
#!/usr/bin/env python3
"""
CP5200 LED Display Library Python Test Suite
Tests all major functions using the Python wrapper
"""

import sys
import time
import os
import tempfile
from typing import List

# Import the CP5200 wrapper
try:
    from cp5200_wrapper import CP5200Display, create_display, CP5200Error
except ImportError:
    print("Error: cp5200_wrapper.py not found. Please run the installation script first.")
    sys.exit(1)

class CP5200Tester:
    def __init__(self, ip_address: str = "192.168.1.100", port: int = 5200, debug: bool = True):
        """Initialize the CP5200 tester
        
        Args:
            ip_address: IP address of the CP5200 controller
            port: Network port
            debug: Enable debug mode
        """
        self.ip_address = ip_address
        self.port = port
        self.debug = debug
        self.display = None
        self.tests_passed = 0
        self.tests_failed = 0
        
    def print_test_result(self, test_name: str, result: int, expected: int = 0):
        """Print test result with formatting"""
        if result == expected:
            print(f"✓ {test_name}: PASSED")
            self.tests_passed += 1
        else:
            print(f"✗ {test_name}: FAILED (error code: {result})")
            self.tests_failed += 1
            
    def print_summary(self):
        """Print test summary"""
        print("\n" + "=" * 50)
        print("TEST SUMMARY")
        print("=" * 50)
        print(f"Tests passed: {self.tests_passed}")
        print(f"Tests failed: {self.tests_failed}")
        print(f"Total tests: {self.tests_passed + self.tests_failed}")
        print("=" * 50)
        
    def setup_display(self):
        """Setup the display connection"""
        print("Setting up CP5200 display connection...")
        try:
            self.display = create_display(self.ip_address, self.port, self.debug)
            print(f"Connected to CP5200 at {self.ip_address}:{self.port}")
            return True
        except CP5200Error as e:
            print(f"Failed to connect to CP5200: {e}")
            return False
            
    def test_version(self):
        """Test library version"""
        print("\n--- Testing Library Version ---")
        try:
            version = self.display.get_version()
            print(f"Library version: {version}")
            self.print_test_result("Get Version", 0)
        except Exception as e:
            print(f"Version test failed: {e}")
            self.print_test_result("Get Version", -1)
            
    def test_network_mode(self):
        """Test network mode configuration"""
        print("\n--- Testing Network Mode ---")
        try:
            self.display.set_network_mode(self.ip_address, self.port)
            self.print_test_result("Set Network Mode", 0)
        except Exception as e:
            print(f"Network mode test failed: {e}")
            self.print_test_result("Set Network Mode", -1)
            
    def test_serial_mode(self):
        """Test serial mode configuration"""
        print("\n--- Testing Serial Mode ---")
        try:
            self.display.set_serial_mode("/dev/ttyAMA0", 115200)
            self.print_test_result("Set Serial Mode", 0)
            # Switch back to network mode for other tests
            self.display.set_network_mode(self.ip_address, self.port)
        except Exception as e:
            print(f"Serial mode test failed: {e}")
            self.print_test_result("Set Serial Mode", -1)
            
    def test_time_sync(self):
        """Test time synchronization"""
        print("\n--- Testing Time Synchronization ---")
        try:
            result = self.display.sync_time()
            self.print_test_result("Time Sync", result)
        except Exception as e:
            print(f"Time sync test failed: {e}")
            self.print_test_result("Time Sync", -1)
            
    def test_brightness_control(self):
        """Test brightness control"""
        print("\n--- Testing Brightness Control ---")
        try:
            # Get current brightness
            current = self.display.get_brightness()
            print(f"Current brightness: {current}")
            
            # Set brightness to 50%
            result = self.display.set_brightness(15)
            self.print_test_result("Set Brightness", result)
            
            # Set auto brightness
            result = self.display.set_brightness(255)
            self.print_test_result("Set Auto Brightness", result)
            
        except Exception as e:
            print(f"Brightness control test failed: {e}")
            self.print_test_result("Brightness Control", -1)
            
    def test_text_display(self):
        """Test text display functionality"""
        print("\n--- Testing Text Display ---")
        
        test_cases = [
            ("Hello CP5200!", 1, 16, 5, 1, 10, 0),
            ("Test Message", 2, 20, 3, 2, 15, 1),
            ("CP5200 Python", 3, 18, 4, 1, 12, 2),
        ]
        
        for i, (text, color, font_size, speed, effect, stay_time, align) in enumerate(test_cases):
            try:
                result = self.display.send_text(0, text, color, font_size, speed, effect, stay_time, align)
                self.print_test_result(f"Send Text {i+1}", result)
                time.sleep(1)  # Wait between messages
            except Exception as e:
                print(f"Text display test {i+1} failed: {e}")
                self.print_test_result(f"Send Text {i+1}", -1)
                
    def test_picture_display(self):
        """Test picture display functionality"""
        print("\n--- Testing Picture Display ---")
        
        # Create a minimal test GIF file
        try:
            with tempfile.NamedTemporaryFile(suffix='.gif', delete=False) as temp_file:
                # Minimal GIF header for testing
                gif_data = bytes([
                    0x47, 0x49, 0x46, 0x38, 0x39, 0x61,  # GIF89a
                    0x01, 0x00, 0x01, 0x00, 0x80, 0x00, 0x00,  # 1x1 pixel
                    0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x21, 0xF9, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2C, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x01, 0x00, 0x00, 0x02, 0x02, 0x44, 0x01, 0x00, 0x3B
                ])
                temp_file.write(gif_data)
                temp_file_path = temp_file.name
                
            result = self.display.send_picture(0, 0, 0, temp_file_path, 5, 1, 10)
            self.print_test_result("Send Picture", result)
            
            # Clean up
            os.unlink(temp_file_path)
            
        except Exception as e:
            print(f"Picture display test failed: {e}")
            self.print_test_result("Send Picture", -1)
            
    def test_clock_display(self):
        """Test clock display functionality"""
        print("\n--- Testing Clock Display ---")
        
        try:
            # Format array: multiline display
            format_array = [1, 0, 0, 0, 0, 0, 0, 0]
            
            # Content array: show date and time
            content_array = [1, 1, 1, 0, 0, 0, 0, 0]
            
            # Colors: white
            colors = [255, 255, 255]
            
            result = self.display.send_clock(0, 30, 0, format_array, content_array, "Current Time", colors, 16)
            self.print_test_result("Send Clock", result)
            
        except Exception as e:
            print(f"Clock display test failed: {e}")
            self.print_test_result("Send Clock", -1)
            
    def test_error_handling(self):
        """Test error handling"""
        print("\n--- Testing Error Handling ---")
        
        # Test with invalid parameters
        try:
            result = self.display.send_text(-1, "Invalid Window", 1, 16, 5, 1, 10, 0)
            print(f"Invalid window test result: {result} (expected error)")
        except Exception as e:
            print(f"Invalid window test exception: {e}")
            
        try:
            result = self.display.send_text(0, "", 1, 16, 5, 1, 10, 0)
            print(f"Empty text test result: {result}")
        except Exception as e:
            print(f"Empty text test exception: {e}")
            
        self.print_test_result("Error Handling", 0)  # Always pass as we're testing error conditions
        
    def test_performance(self):
        """Test performance with multiple messages"""
        print("\n--- Testing Performance ---")
        
        start_time = time.time()
        
        try:
            # Send multiple text messages
            for i in range(10):
                message = f"Performance Test {i}"
                result = self.display.send_text(i % 4, message, 1, 16, 5, 1, 5, 0)
                time.sleep(0.1)  # 100ms delay
                
            end_time = time.time()
            duration = end_time - start_time
            
            print(f"Performance test completed in {duration:.3f} seconds")
            self.print_test_result("Performance", 0)
            
        except Exception as e:
            print(f"Performance test failed: {e}")
            self.print_test_result("Performance", -1)
            
    def test_communication_modes(self):
        """Test different communication modes"""
        print("\n--- Testing Communication Modes ---")
        
        try:
            # Test network mode
            self.display.set_network_mode(self.ip_address, self.port)
            print("Switched to network mode")
            
            # Test serial mode
            self.display.set_serial_mode("/dev/ttyAMA0", 115200)
            print("Switched to serial mode")
            
            # Switch back to network mode
            self.display.set_network_mode(self.ip_address, self.port)
            print("Switched back to network mode")
            
            self.print_test_result("Communication Modes", 0)
            
        except Exception as e:
            print(f"Communication modes test failed: {e}")
            self.print_test_result("Communication Modes", -1)
            
    def run_all_tests(self):
        """Run all tests"""
        print("CP5200 LED Display Library Python Test Suite")
        print("=" * 50)
        print("This test suite will test all major functions of the CP5200 library")
        print("Make sure your CP5200 display controller is connected and configured")
        print("=" * 50)
        
        # Setup display connection
        if not self.setup_display():
            print("Failed to setup display connection. Exiting.")
            return False
            
        # Run all tests
        self.test_version()
        self.test_network_mode()
        self.test_serial_mode()
        self.test_time_sync()
        self.test_brightness_control()
        self.test_text_display()
        self.test_picture_display()
        self.test_clock_display()
        self.test_error_handling()
        self.test_performance()
        self.test_communication_modes()
        
        # Print summary
        self.print_summary()
        
        # Return success/failure
        if self.tests_failed == 0:
            print("All tests passed! CP5200 library is working correctly.")
            return True
        else:
            print("Some tests failed. Please check your configuration.")
            return False

def main():
    """Main entry point"""
    import argparse
    
    parser = argparse.ArgumentParser(description='CP5200 LED Display Library Test Suite')
    parser.add_argument('--ip', default='192.168.1.100', help='IP address of CP5200 controller')
    parser.add_argument('--port', type=int, default=5200, help='Network port')
    parser.add_argument('--no-debug', action='store_true', help='Disable debug mode')
    
    args = parser.parse_args()
    
    # Create tester and run tests
    tester = CP5200Tester(args.ip, args.port, not args.no_debug)
    success = tester.run_all_tests()
    
    sys.exit(0 if success else 1)

if __name__ == "__main__":
    main()

File: cp5200/cp5200_test.c
/*
 * CP5200 LED Display Library Test Suite
 * Tests all major functions of the CP5200 library
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <cp5200.h>

// Test configuration
#define TEST_IP_ADDRESS "192.168.1.100"
#define TEST_PORT 5200
#define TEST_SERIAL_PORT "/dev/ttyAMA0"
#define TEST_BAUD_RATE 115200

// Test results
int tests_passed = 0;
int tests_failed = 0;

void print_test_result(const char* test_name, int result) {
    if (result == 0) {
        printf("✓ %s: PASSED\n", test_name);
        tests_passed++;
    } else {
        printf("✗ %s: FAILED (error code: %d)\n", test_name, result);
        tests_failed++;
    }
}

void print_summary() {
    printf("\n" "=" * 50 "\n");
    printf("TEST SUMMARY\n");
    printf("=" * 50 "\n");
    printf("Tests passed: %d\n", tests_passed);
    printf("Tests failed: %d\n", tests_failed);
    printf("Total tests: %d\n", tests_passed + tests_failed);
    printf("=" * 50 "\n");
}

// Test 1: Library version
int test_version() {
    printf("\n--- Testing Library Version ---\n");
    _get_cp5200_version();
    return 0; // Version function doesn't return error codes
}

// Test 2: Debug mode
int test_debug_mode() {
    printf("\n--- Testing Debug Mode ---\n");
    _set_cp5200_debug();
    return 0;
}

// Test 3: Network communication setup
int test_network_setup() {
    printf("\n--- Testing Network Setup ---\n");
    _set_cp5200_ipcomm(TEST_IP_ADDRESS, TEST_PORT);
    _set_cp5200_send_mode(0); // TCP/IP mode
    return 0;
}

// Test 4: Serial communication setup
int test_serial_setup() {
    printf("\n--- Testing Serial Setup ---\n");
    _set_cp5200_rs232comm(TEST_SERIAL_PORT, TEST_BAUD_RATE);
    _set_cp5200_send_mode(1); // RS-232 mode
    return 0;
}

// Test 5: Time synchronization
int test_sync_time() {
    printf("\n--- Testing Time Synchronization ---\n");
    int result = SyncTime();
    print_test_result("Time Sync", result);
    return result;
}

// Test 6: Brightness control
int test_brightness_control() {
    printf("\n--- Testing Brightness Control ---\n");
    
    // Test getting current brightness
    int current_brightness = BrightnessControl(0, 0);
    printf("Current brightness: %d\n", current_brightness);
    
    // Test setting brightness to 50%
    int result = BrightnessControl(1, 15);
    print_test_result("Set Brightness", result);
    
    // Test setting auto brightness
    result = BrightnessControl(1, 255);
    print_test_result("Set Auto Brightness", result);
    
    return result;
}

// Test 7: Text display
int test_text_display() {
    printf("\n--- Testing Text Display ---\n");
    
    // Test basic text
    int result = SendText(0, "Hello CP5200!", 1, 16, 5, 1, 10, 0);
    print_test_result("Send Text - Basic", result);
    
    // Test with different parameters
    result = SendText(1, "Test Message", 2, 20, 3, 2, 15, 1);
    print_test_result("Send Text - Advanced", result);
    
    // Test UTF-8 text (if supported)
    result = SendText(2, "CP5200 Test", 3, 18, 4, 1, 12, 2);
    print_test_result("Send Text - UTF8", result);
    
    return result;
}

// Test 8: Picture display
int test_picture_display() {
    printf("\n--- Testing Picture Display ---\n");
    
    // Test with a sample GIF file (create a dummy file for testing)
    FILE* test_file = fopen("test_image.gif", "wb");
    if (test_file) {
        // Write minimal GIF header
        unsigned char gif_header[] = {
            0x47, 0x49, 0x46, 0x38, 0x39, 0x61, // GIF89a
            0x01, 0x00, 0x01, 0x00, 0x80, 0x00, 0x00, // 1x1 pixel
            0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x21, 0xF9, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2C, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x01, 0x00, 0x00, 0x02, 0x02, 0x44, 0x01, 0x00, 0x3B
        };
        fwrite(gif_header, 1, sizeof(gif_header), test_file);
        fclose(test_file);
        
        int result = SendPicture(0, 0, 0, "test_image.gif", 5, 1, 10);
        print_test_result("Send Picture", result);
        
        // Clean up test file
        unlink("test_image.gif");
        return result;
    } else {
        printf("Could not create test image file\n");
        return -1;
    }
}

// Test 9: Clock display
int test_clock_display() {
    printf("\n--- Testing Clock Display ---\n");
    
    // Format array: multiline display
    int format[] = {1, 0, 0, 0, 0, 0, 0, 0};
    
    // Content array: show date and time
    int content[] = {1, 1, 1, 0, 0, 0, 0, 0};
    
    // Colors: white
    int colors[] = {255, 255, 255};
    
    int result = SendClock(0, 30, 0, format, content, "Current Time", colors, 16);
    print_test_result("Send Clock", result);
    
    return result;
}

// Test 10: Window splitting
int test_window_splitting() {
    printf("\n--- Testing Window Splitting ---\n");
    
    // Test window configuration
    int window_config[] = {0, 1, 2, 3};
    int result = SplitWindow(0, window_config, 4);
    print_test_result("Split Window", result);
    
    return result;
}

// Test 11: Communication mode switching
int test_communication_modes() {
    printf("\n--- Testing Communication Modes ---\n");
    
    // Test TCP/IP mode
    _set_cp5200_send_mode(0);
    printf("Switched to TCP/IP mode\n");
    
    // Test RS-232 mode
    _set_cp5200_send_mode(1);
    printf("Switched to RS-232 mode\n");
    
    // Test RS-485 mode
    _set_cp5200_send_mode(2);
    printf("Switched to RS-485 mode\n");
    
    return 0;
}

// Test 12: Error handling
int test_error_handling() {
    printf("\n--- Testing Error Handling ---\n");
    
    // Test with invalid parameters
    int result = SendText(-1, "Invalid Window", 1, 16, 5, 1, 10, 0);
    printf("Invalid window test result: %d (expected error)\n", result);
    
    result = SendText(0, "", 1, 16, 5, 1, 10, 0);
    printf("Empty text test result: %d\n", result);
    
    return 0;
}

// Test 13: Performance test
int test_performance() {
    printf("\n--- Testing Performance ---\n");
    
    clock_t start, end;
    double cpu_time_used;
    
    start = clock();
    
    // Send multiple text messages
    for (int i = 0; i < 10; i++) {
        char message[50];
        sprintf(message, "Performance Test %d", i);
        SendText(i % 4, message, 1, 16, 5, 1, 5, 0);
        usleep(100000); // 100ms delay
    }
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    
    printf("Performance test completed in %.3f seconds\n", cpu_time_used);
    return 0;
}

// Main test runner
int main() {
    printf("CP5200 LED Display Library Test Suite\n");
    printf("=" * 50 "\n");
    printf("This test suite will test all major functions of the CP5200 library\n");
    printf("Make sure your CP5200 display controller is connected and configured\n");
    printf("=" * 50 "\n");
    
    // Run all tests
    test_version();
    test_debug_mode();
    test_network_setup();
    test_serial_setup();
    test_sync_time();
    test_brightness_control();
    test_text_display();
    test_picture_display();
    test_clock_display();
    test_window_splitting();
    test_communication_modes();
    test_error_handling();
    test_performance();
    
    // Print summary
    print_summary();
    
    // Return appropriate exit code
    if (tests_failed == 0) {
        printf("All tests passed! CP5200 library is working correctly.\n");
        return 0;
    } else {
        printf("Some tests failed. Please check your configuration.\n");
        return 1;
    }
}
